### PR TITLE
WIP: Use declarative Aframe code with imperative hooks instead of Aframe-React

### LINKED ISSUE #7: Game performance (framerate) is very slow
I suspect this is from React constantly recreating DOM elements that are then picked up by Aframe. Needs investigation.

To start, the [Aframe-React README.md](https://github.com/supermedium/aframe-react) starts with this lovely message:

> I recommend using vanilla A-Frame and [aframe-state-component](https://npmjs.org/aframe-state-component) with [static templating](https://mozilla.github.io/nunjucks/) over `aframe-react`. React wastes a lot of cycles and incurs a lot of memory garbage. aframe-react is often abused where it is too easy to place 3D/real-time logic at the React layer, causing poor performance (e.g., doing React renders on ticks). aframe-react applications frequently ignore the prescribed ECS framework of A-Frame. Internally, React does tons of computation to compute what changed, and flushes it to the entire application. It is apparent React ecosystem does not care much about memory as most examples allocate functions and objects in the render method, and where immutables are popular. With only ~10ms per frame to do all computation, there is little room for React's massive system.

It would be difficult for us to use `aframe-state-component` as our state is generally stored in React. I think a better path forward would be to dial back on the React-isms in our Rooms code and do direct DOM manipulations on previously rendered static HTML where possible.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## Summary

Fixes #7.

There are two major interlinked architectural changes in this PR:

* Aframe-React is replaced with a raw `template.html` placed in the page's document and hooks that set up events on those DOM elements / Aframe entities.
* Socket & room context initialization are moved to a `@connection` directory (`src/connection/`) and restructured so that room initialization is only started after the server sends a snapshot of room data.

See the revamped `.md` docs for more details.

The latter change makes it so there's no longer a default state applied locally while waiting for an update from the server. That's a nice little win, I think!

### Followups

A few nice-to-have improvements are a bit too high complexity / low reward for this PR in my opinion but could be filed as followup PRs:

* Throttling `currentTime` updates to the local `roomData.currentTime` state -both for the client sending it and other clients receiving it- to minimize React re-renders 
  * In general, can we _not_ trigger a rerender of most of the reactive room UI whenever a button is pressed...? It's not affecting my computer but still feels unfortunate. _(I miss coding in MobX...)_
* Calculating the lag time between receiving the video currentTime and when it actually starts playing, then increasing the set `currentTime` by that much too (in addition to the `videoSyncInterval / 2000`).
* Using `KaraokeEvent.UsernameSet` as the send event instead of the first `KaraokeEvent.RoomDataHydration`, to get the username sent slightly sooner & with one less socket message.

### CODE REVIEW

File: public/js/dynamic-room.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  init: function () {
+    var { room } = this.getUrlParams();
+
+    this.el.setAttribute("networked-scene", {
+      adapter: "webrtc",
+      audio: "true",
+      room,
+      serverURL: `:3001`,
+    });
+  },
```
Comment (JoshuaKGoldberg) [inline public/js/dynamic-room.js]: Todo: move this into a `useEffect` hook so we can delete `getUrlParams`...

File: public/js/video-sync.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+// Put this on the a-video tag in the template
+// This will run for everybody in the room
+
+AFRAME.registerComponent("video-sync", {
```
Comment (JoshuaKGoldberg) [inline public/js/video-sync.js]: Todo: is not using this why the video isn't showing...?

File: src/pages/_document.tsx [hunk 3]
```diff
@@ -1,10 +1,14 @@
...
+import React from "react";
+
+import { RoomBody } from "./room/RoomBody";
 
 class MyDocument extends Document {
   render() {
     return (
       <Html>
         <Head>
+          {/* Todo: asynchronously load these if in the room page */}
```
Comment (JoshuaKGoldberg) [inline src/pages/_document.tsx]: Todo: these todos 😄

File: src/pages/room/DynamicScene/index.tsx [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+export const DynamicScene: React.FC = () => {
+  const emit = useRoomConnection();
+
+  return (
+    <EmitContext.Provider value={emit}>
+      <DynamicSceneEvents />
+    </EmitContext.Provider>
+  );
+};
```
Comment (JoshuaKGoldberg) [inline src/pages/room/DynamicScene/index.tsx]: todo: explain in docs how this new architecture is set up:

* DynamicSceneEvents and the contexts
* `useOnClick` and paying attention to elements
   * File issue for making the roomContext provider not constantly change & cause rerenders
   * Explain why setting & emitting info is local / onClick compared to the useEffect for applying dom changes

File: src/pages/room/RoomBody/RawHtml.tsx [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import React from "react";
+
+// https://github.com/networked-aframe/networked-aframe/issues/226
+// Ugh.
```
Comment (JoshuaKGoldberg) [inline src/pages/room/RoomBody/RawHtml.tsx]: Todo: maybe update this to explain the newer usage reasons?

File: src/pages/room/RoomContext.ts [hunk 6]
```diff
@@ -13,10 +13,15 @@ export const RoomContext = React.createContext<RoomContextValue>(null!);
 const useGetterAndSetter = <Value>(value: Value) => {
   const [stateValue, set] = useState(value);
 
-  return {
-    get: () => stateValue,
-    set,
-  };
+  const values = React.useMemo(
```
Comment (JoshuaKGoldberg) [inline src/pages/room/RoomContext.ts]: Todo: pretty sure this isn't doing what I want (see earlier todo about not causing rerenders)... I think useSt8 might be the answer.

**Review by fedGL — APPROVED** [review]
LGTM

**Pull request merged**